[Qol]Fixit nested static value

gibachan opened the pull request:
<!-- What's in this pull request? -->
This PR fixes a insufficient fixit on use of static value in nested structure.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves [SR-5715](https://bugs.swift.org/browse/SR-5715).

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

gibachan committed:
[Qol]Add a test for nested static value

Variable name in nested static value should be fixed to have complete declarelation

gibachan committed:
[Qol]Use complete declaration

CodaFi commented on lib/Sema/CSDiag.cpp:
```diff
@@ -2559,7 +2559,7 @@ diagnoseTypeMemberOnInstanceLookup(Type baseObjTy,
   SmallString<32> typeName;
   llvm::raw_svector_ostream typeNameStream(typeName);
-  typeNameStream << nominal->getName() << ".";
+  typeNameStream << nominal->getDeclaredInterfaceType() << ".";
```
I'm a little hesitant to print this directly. but it seems to be correct.  Could you add a test that uses a generic nested type as well?

slavapestov replied:
The declared interface type should be fine. It will use sugared generic parameter names and include the full parent type. So for a nested generic type, it should print `Outer<T>.Inner<U, V>`, etc, and we should add a test for this case.

slavapestov replied:
Actually to handle members of protocols you want getSelfInterfaceType() instead. This is identical to getDeclaredInterfaceType() for concrete types, but for protocols it has 'Self' instead of the existential.

Right now, this produces a bad fixit:
```
protocol P {}

extension P {
  static func f() {}

  func g() {
    f
  }
}

ext.swift:7:5: error: static member 'f' cannot be used on instance of type 'Self'
    f
    ^
    P.

```

Note that the correct fix is to add 'Self.', not 'P.'.

@gibachan Do you mind adding a test for this case as well?

gibachan replied:
Thank you for kind response.
I made those fixes, but now I get a strange error on the test you advised above.
When I run the test by `utils/run-test`, the error message says "use of unresolved identifier 'f'" instead of "static member 'f' cannot be used ...". However if I run it by `../build/Ninja-ReleaseAssert/swift-macosx-x86_64/bin/swift`, it says "static member 'f' cannot be used ..." as I expect. I do not know the difference between them. Am I missing something?

xwu commented on test/expr/postfix/dot/dot_keywords.swift:
```diff
@@ -77,3 +77,19 @@ class ClassWithDeinitMember {
+    func f() -> Int {
+        return x  // expected-error {{static member 'x' cannot be used on instance of type 'SR5715.A.B'}} {{16-16=SR5715.A.B.}}
+    }
+}
```
Drive-by nit: could you please reformat these lines to use two-space indents like the rest of the file, and restore the trailing newline?

gibachan replied:
Thanks, I did reformat lines.

gibachan committed:
Fix spaces

slavapestov commented:
Hi @gibachan, can you take a look at this PR? https://github.com/apple/swift/pull/11609 It looks like there are two very similar code paths in diagnostics to handle the same error of an instance member being used on static type value, or vice versa. Maybe there's a function that can be factored out here.

cc @Kacper20

gibachan committed:
Use getSelfInterfaceType()

getDeclaredInterfaceType() is invalid for protocol. getSelfInterfaceType() is good for both concrete types and protocol.

gibachan committed:
Add a test for protocol with static member

gibachan committed:
Add a test for protocol with static member

gibachan committed:
Restore trailing newline

slavapestov commented:
Looks good, thanks! I'll look at the other code path later, maybe they're not really possible to unify.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
Hi @gibachan, can you take a look at the test failure?

```
/Users/buildnode/jenkins/workspace/swift-PR-osx-smoke-test/branch-master/swift/test/expr/postfix/dot/dot_keywords.swift:104:31: error: incorrect message found
    f()   // expected-error {{static member 'f' cannot be used on instance of type 'Self'}} {{12-12=Self.}}
                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                              use of unresolved identifier 'f'
```

gibachan committed:
Fix the position of fixit

gibachan commented:
Hi @slavapestov, thank you for your support!
The failed test expects that statement lacks of 'Self.' but it was different because of swift version.
The expected error happens only with Swift4. With Swift3, it would be as described above.
Would you please let me know how to handle the different results by Swift version in test case like this situation? or any better approach?

slavapestov commented:
Oh right, Swift 3 had a strange quirk with unqualified lookup vs static members...

You could update the test's RUN line to pass -swift-version 4, and add a new test case to test/Compatibility/ to test the old behavior.

Also do you know how to run the tests locally? That would save you a lot of time while iterating on this.

gibachan committed:
Use Swift4

gibachan committed:
Add test for compatibility

gibachan commented:
Thanks @slavapestov! I followed your advice and I tried tests locally with `utils/build-script -RT` then it passed without any unexpected failures. 
Could you please take a look at it?

slavapestov commented:
@swift-ci Please smoke test
